atacworks/dl4atac/utils.py

```diff
@@ -399,3 +416,137 @@ def log(self, names, normalizer=1.0, reset=True):
             string += ' | {}: {:.2f}'.format(name, elapsed_time)
         myprint(string)
+
+
+def _get_tiling_intervals(sizes, intervalsize):
+    """Produce intervals of given chromosomes.
+
+    Tile from start to end of given chromosomes, shifting by given length.
+
+    Args:
+        sizes: Pandas df containing columns 'chrom' and 'length',
+            with name and length of required chromosomes
+        intervalsize: length of intervals
+        shift: distance between starts of successive intervals.
+
+    Returns:
+        Pandas DataFrame containing chrom, start, and end of tiling intervals.
+
+    """
+    # Default: non-overlapping intervals
+    shift = intervalsize
+
+    # Create empty DataFrame
+    intervals = pd.DataFrame()
+
+    # Create intervals per chromosome
+    for i in range(len(sizes)):
+        chrom = sizes.iloc[i, 0]
+        chrend = sizes.iloc[i, 1]
+        starts = range(0, chrend - (intervalsize + 1), shift)
+        ends = [x + intervalsize for x in starts]
+        intervals = intervals.append(pd.DataFrame(
+            {'chrom': chrom, 'start': starts, 'end': ends}))
+
+    # Eliminate intervals that extend beyond chromosome size
+    intervals = intervals.merge(sizes, on='chrom')
+    intervals = intervals[intervals['end'] < intervals['length']]
+
+    return intervals.loc[:, ('chrom', 'start', 'end')]
+
+
+def get_intervals(sizesfile, intervalsize, out_dir, wg=None, val=None,
+                  holdout=None, nonpeak=None, peakfile=None):
+    """Read chromosome sizes and generate intervals.
+
+     Args:
+         sizesfile: BED file containing sizes of each chromosome.
+         intervalsize: Size of the intervals at each row.
+         out_dir: Directory to save the output files to.
+         wg: Create intervals for the whole genome.
+         val: Chromosome to reserve for validation.
+         holdout: Chromosome to reserve for evaluation.
+         nonpeak: Ratio of nonpeak to peak intervals desired in training
+         dataset.
+         peakfile: File with clean peaks to know which intervals have non-zero
+         values. Only useful if nonpeak is greater than one.
+
+    Returns:
+         Paths of files saved.
+
+    """
+    # Read chromosome sizes
+    sizes = read_sizes(sizesfile)
+
+    # Create the output dir
+    if not os.path.exists(out_dir):
+        os.makedirs(out_dir)
+    # Generate intervals
+    if wg:
+
+        # Generate intervals tiling across all chromosomes in the sizes file
+        _logger.info("Generating intervals tiling across all chromosomes \
+            in sizes file: " + sizesfile)
+        intervals = _get_tiling_intervals(sizes, intervalsize)
+
+        # Write to file
+        out_file_name = str(intervalsize) + '.genome_intervals.bed'
+        wg_file_path = os.path.join(out_dir, out_file_name)
+        df_to_bed(intervals, wg_file_path)
+        _logger.info('Done!')
+        return wg_file_path
+
+    elif not (val is None or holdout is None):
```

Comment:
So.. if the only one of val or holdout are provided, this function won't do anything?
We should check whether both val and holdout are provided and throw an error if only one is given.

Reply:
added a check in scripts/cmd_args.py. If args.mode == train, then it will require both val and holdout chroms.